testing/preact.js

```diff
@@ -0,0 +1,69 @@
+import * as preact from /*OK*/ 'preact';
+
+/**
+ * This file introduces a helper for draining Preact's queue of renders and effects to run.
+ * We use this as part of the afterEach() cleanup in unit tests, to ensure no effects are run
+ * subsequent tests.
+ *
+ * There is still a test isolation issue in that an effect can asynchronously schedule work
+ * which cannot be guarded from at this layer. For that we'd likely need to refresh the window
+ * in between each test run.
+ *
+ * We should be able to remove this file if this feature lands in Preact.
+ * @fileoverview
+ */
+
+const rafs = [];
+/**
+ * @param {*} cb
+ * @return {number}
+ */
+function flushableRaf(cb) {
+  cb.__completed = false;
+  rafs.push(cb);
+  return requestAnimationFrame((ts) => {
+    if (cb.__completed) {
+      return;
+    }
+    cb.__completed = true;
+    cb(ts);
+  });
+}
+
+let flushRender;
+
+/**
+ * Set to ensure a render is only called once.
+ * @type {WeakSet<() => void>}
+ */
+const flushedRenderSet = new WeakSet();
+
+/**
+ * @param {() => voic} process
+ */
+function flushableRender(process) {
+  flushRender = process;
+  Promise.resolve().then(() => {
+    if (flushedRenderSet.has(process)) {
+      return;
+    }
+    flushedRenderSet.add(process);
+    flushRender = null;
+    return process();
+  });
```

Comment:
This one should actually a boolean, not a weakset. `process` is a module constant in preact, and we need to continually rerun it. But, it may be schedule multiple times before finally running, and we only want to run once.

```suggestion
/**
 * Ensure render is debounced
 */
const debounceRendered = true;

/**
 * @param {() => void} process
 */
function flushableRender(process) {
  flushRender = process;
  debounceRendered = false;
  Promise.resolve().then(() => {
    if (debounceRendered) {
      return;
    }
    debounceRendered = true;
    flushRender = null;
    return process();
  });
```

Reply:
Yes - sorry, this was the reason for the `while (pending)` in my Gist that you asked about (just saw the comment now):
https://gist.github.com/developit/3d7a69a2e819c8d11888f3a310a98f8f#gistcomment-3921021

Any time the commit/flush callback from `debounceRendering` is invoked, all of the enqueued components are rendered. If a component enqueues additional updates during that render (while the queue is draining, but during/after rendering of that component), it will be enqueued in a second render pass that gets deferred by calling `options.debounceRendering`.

Here's an annotated version of my gist suggestion that better illustrates:

```js
// promise defined when any renders are enqueued:
// (resolves when the queue is flushed, undefined when idle)
let pending;

options.debounceRendering = f => {
  // the render queue now contains work (from setState, etc)
  pending = Promise.resolve().then(() => {
    // flush is synchronous, so we have to reset the indicator here to avoid
    // clobbering a new `pending` set by a nested debounceRendering call:
    pending = undefined;
    // flush all *currently* enqueued rendering:
    f();
  });
};

export async function flush() {
  // invoke any pending effects first, since they may enqueue renders (but cannot enqueue effects):
  await flushRaf();
  // if any renders were enqueued...
  while (pending) {
    await pending;  // ...wait for them to finish (just a Promise tick)
    await flushRaf();  // ...then invoke any effects enqueued by the render
  }
}
```